Hope fork done

slwoodward opened the pull request:
Game choice now set at server and sent to clients.  Changed start-up dialog so default server doesn't need to change name when port# is changed.
Some minor bug fixes.
Incorporated user feedback-- especially regarding presentation of user runs (long runs became illegible when there were 6 players).

slwoodward committed:
Project begun

Modifying Hand and Foot code so that it supports Liverpool.  Began work on modifying ruleset.   Added ability to sort by suit.

slwoodward committed:
Prepare cards button per group (#1)

* Buttons to assign cards created.
  - Rather than single prepare_cards_btn now have assign_cards_btns.  
Also created method to change text within a button (probably will not use this)
  - Assignment buttons now in dictionary. Key (a tuple) designates player and which button in the list.

* Implemented more sorting options: by suit and Aces high / low sorting options.

* Pick up pile is working (have not yet begun work on buying cards feature)
  - Liverpool doesn't require that top card be played--modifications made to support this.

* common/Liverpool.py work has begun.

slwoodward committed:
Played cards dictionary and display (#3)


Ran into 2 related problems, both traceable to the fact that Liverpool allows you to play on others sets and groups.
1. played_cards in HandAndFoot can be totally local, but it cannot be local in Liverpool.
2. indexing scheme needs to be consistent throughout -- instead of keying on card.number, key on tuple = player#, group#.   
After some false starts those issues are resolved.
--
Changes in this commit pass more variables from RunClient to handView. 


Now merge visible cards which has a dictionary for each player to one big dictionary, so can play on existing groups. 
 Note this  made  compressGroups in tableView more complicated, but will simplifies other stuff.

There are some (many) bugs (especially in TableView) and many rules still need to be written. BUT:
- coarse display in place
- you can play cards on any players group.
- HandAndFoot still works.
-pyinstaller still works with HandAndFoot.

slwoodward committed:
Fix known bugs (#4)

* Made progress, many bugs fixed.  
* Stopped at Server question that affects checking if have melded in 2nd round.

slwoodward committed:
Fixed visible runs in Liverpool (#6)

* started writing code to check runs

* visible_cards needed definition modification in Liverpool; 
In Liverpool (but not HandAndFoot) changed structure of visible_cards so that it is now a list containing a single dictionary, instead of a dictionary for each player.  That single dictionary is the the one with the most cards played of all the dictionaries received from the player channels.  This allows you to preserve cards played by a player who later drops out. AND to preserve the ordering of runs, which should simplify assigning wilds.

slwoodward committed:
Runs rules and sorting (#7)

* Working on Rules for Runs
Still get exception when try to play a joker in a run (no problem in sets).  This affects my ability to test joker behavior.
Exception only raised when hit play prepared cards button.
-- 
To do rule checking have to preserve  values assigned to Jokers.
-> gave card another attribute (not used by server) -- tempnumber.  Need this to preserve assigments for Jokers, and whether Aces are high or low. 

Created method proccessRuns  in liverpool.py.  Puts run in order and  does some of the rule checking.  
Currently call it more than necessary because call it on all runs -- even those that are not being added to, and call 
it before and during canPlayGroup.  
However, -- rule checking for Runs seems be working and have Aces hi/lo but not both working.  
--fine tune flow after figure out wild card bug.

AAlso still to be done:
Need to update meld_threshold buttons after OK is clicked by all -- else test rounds runs are displayed in set format.

slwoodward committed:
Update Liverpool.py

Modified canMeld

slwoodward committed:
Runs rules and sorting (#8)

Lots of rule checking. However, there is now a bug that appears if prepare a wild for a RUN and then hit play prepared cards.

Significant changes-- Cards now have property card.tempnumber so that can track assigned values of wilds, and when Aces are high. Server code does NOT use this property, as it is not passed between players and server.

processRuns is a new method that puts cards in order (including unassigned jokers). It simultaneously does rules checking.
Does not yet handle jokers that can be either High or Low. Need to sort out whether this should be called before canPlayGroup or inside it. (currently does both)..  Note canMeld assumes that it will be called within canPlayGroup.
However, need to make certain that processed cards are what ends up transmitted to the server.

slwoodward committed:
Merge branch 'liverpool' of https://github.com/slwoodward/Remote-Card-Games into liverpool

slwoodward committed:
Oct28 cleanup (#9)

This now supports Liverpool with some caveats.  Cannot yet buy cards, and cannot designate where Aces and wilds should be when they're at the ends of runs.   Also this needs more polishing., and more thorough testing.

You can play on other players melds.  To support this needed to send visible cards to controller. Changed structure so that visible cards is now a list containing at most one dictionary, and the keys are tuples (player no. , group no. ).  There is a button for each group which is used when preparing cards.
Mechanics a bit different because you designate the group cards go to explicitly upon preparing them (each group has a different button).

I double checked that pyinstaller works.

slwoodward committed:
Merge branch 'liverpool' of https://github.com/slwoodward/Remote-Card-Games into liverpool

slwoodward committed:
Buy option (#10)

Concluded buyOption had to be controlled by server.
Got server to send announcement to eligible player that they could buy top card.
Server now lets eligible clients know when buying opportunity occurs.
Clients let server know if they want the card.    Gives purchased card and top card to player who buys it.
Reset want_card so that don't accidentally purchase future discard if wanted earlier one.
--Other new features:
 Suit Symbols
changed font so In Liverpool: Tableview now displaying suit symbols.
Players can join between rounds  in Liverpool.
Moved message that appears at start and between rounds to ruleset (both HandAndFoot and Liverpool).

Moved processRuns outside of ruleset and into RunManagement.py.  
Updated structure so that now have controller handling prepared cards (good).   

* began work on wildsHiLo method, but decided to put that in new Branch, so commented out that work here.
Plan is to make branch off of this focused on wildHiLo, and then get BuyOption working with ambiguous wilds handled  automatically (by making them high).  Can come back to wildHILo work later.

Confirmed BuyOption working, pyinstaller working, HandAndFoot still working.

slwoodward committed:
clean-up

slwoodward committed:
Fixed rules so can test play with others.

slwoodward committed:
Consolidate methods calledby liverpool buttons (#11)


* fixed disappearing wilds bug plus some clean-up.

slwoodward committed:
Designate ambiguous wilds (#13)

Designating wilds hi or low is working.  Also fixed some bugs in processRuns.

slwoodward committed:
Add Liverpool to this fork of RemoteCardGames. (#12)

Modifying Hand and Foot code so that it supports Liverpool.  

In this version-
Verified:
- HandAndFoot still works.
-pyinstaller still works (creates executable)

Player dropping out still represents a challenge in Liverpool, but can now restart game in any round.

**Shared board -- can play on other players:
* Prepare cards button per group (#1)
  - Rather than single prepare_cards_btn now have assign_cards_btns.    (Also created method to change text within a button, but didn't end up using it).

* visible_cards needed definition modification in Liverpool; 
In Liverpool (but not HandAndFoot) changed structure of visible_cards so that it is now a list containing a single dictionary, instead of a dictionary for each player.  That single dictionary is the the one with the most cards played of all the dictionaries received from the player channels.  This allows you to preserve cards played by a player who later drops out. AND to preserve the ordering of runs, which should simplify assigning wilds.   Instead of keying on card.number, key on tuple = player#, group#.   

*. played_cards in HandAndFoot can be managed  locally, but it cannot be local in Liverpool.
* Mechanics a bit different because you designate the group cards go to explicitly upon preparing them (each group has a different button) but you don't designate wilds value until after you hit play prepared cards button.


**Buying Option:
* Pick up pile is working, as is buying cards option.
  - Liverpool doesn't require that top card be played--modifications made to support this.

* Concluded buyOption had to be controlled by server.
Got server to send announcement to eligible players that they could buy top card, so server lets eligible clients know when buying opportunity occurs.  Clients let server know if they want the card.    Gives purchased card and top card to player who buys it.  Server also  announces if someone buys the card or picks up the pile (or top discard).
Reset want_card so that don't accidentally purchase future discard if wanted earlier one.


Ran into 2 related problems, both traceable to the fact that Liverpool allows you to play on others sets and groups.

**Also have runs:
* Implemented more sorting options: by suit and Aces high / low sorting options.
* To do rule checking have to preserve  values assigned to Jokers.
-> gave card another attribute (not used by server) -- tempnumber.  Need this to preserve assigments for Jokers, and whether Aces are high or low. 
* Created method proccessRuns  in liverpool.py.  Puts run in order and  does some of the rule checking.  

** Other variations in rules:
Modified canMeld
Moved message that appears at start and between rounds to ruleset (both HandAndFoot and Liverpool).
etc.

slwoodward committed:
Prettify (#22)

Latest version, and has been beta tested.  There are a few cosmetic changes needed and there's still a problem with Liverpool if a player loses their connection.  Current fix to that is to restart game at the appropriate round (game now supports starting in later rounds).

slwoodward committed:
Prettify (#24)

* Enhanced Tableview display in Liverpool
* Some bug fixes 
* pick-ups announced
             If you pick up the pile (HandAndFoot) or the top card (Liverpool)it is now announced.
* Can now start at any round !
           Done on server side. To start now need:
           py RunServer.py server:port Ruleset  # where # = starting round.
Features that were not in Liverpool branch that was just approved for merger into this fork's master:
* refresh draw pile
* cleaned up sortkeys code
* Update TableView.py (replaced 11,12,13,1 with J,Q,K,A in display of runs in Liverpool).
* fixed some bugs in RunManagement
* fixed bug in GamerServer that  caused crashes

Addressed all but 1 comment (tableView imports from rules)  from earlier review of Liverpool branch.

put check on length of visible_cards  (to insure no cards lost) in a separate method and now only call it on update, instead of on every broadcast

slwoodward committed:
Player disconnect liverpool (#27)

If player disconnects Liverpool should recover -- added methods to GameServer to handle this.
Checked that compiled, and that both HandAndFoot and Liverpool are working.  Removed print statements used in debugging.

Also cmade some cosmetic changes in response to user feedback:
Joker not offered for sale at beginning of game
Liverpool sorting buttons are a little bit more separated.
Some messages around card buying have been changed.

slwoodward committed:
Update RunClient.py

Give user 2nd chance to enter correct name of game.

slwoodward committed:
Update RunClient.py

hwoodward committed:
Merge branch 'master' into liverpool

slwoodward committed:
Reconciling.

slwoodward committed:
Server pick game (#29)

Main update is that now the client gets the name of the ruleset from the server.  This required changes to RunClient.py,
Controller.py, Clientstate.py,  GameServer.py and PlayerChannel.py.

Minor change:  hostinfo and port now entered separately.
Did this so that players would not need to type in host name when we switched to another port#, but could continue to use default.  

Also verified the if/while statements used to set player_index in RunClient.py are needed (while loop goes through a couple of times).

slwoodward committed:
Liverpool improvements (#30)

* Rules enforce last round is pat hand.
* Acknowledge response to buying opportunity.
* Wrap groups
                        For Liverpool with lots of players the text in groups needs to wrap around or you can't see the entire run.

- Verified that RunClient.py compiles.
- Removed some errant print statements.
- Added more info to print statements in RunClient.

Created RunClientWithJanServer -- a version that runs with the last releases server (does not get game  choice from server).

slwoodward committed:
Add Liverpool (#28)

* Project begun

Modifying Hand and Foot code so that it supports Liverpool.  Began work on modifying ruleset.   Added ability to sort by suit.

* Prepare cards button per group (#1)

* Buttons to assign cards created.
  - Rather than single prepare_cards_btn now have assign_cards_btns.  
Also created method to change text within a button (probably will not use this)
  - Assignment buttons now in dictionary. Key (a tuple) designates player and which button in the list.

* Implemented more sorting options: by suit and Aces high / low sorting options.


Ran into 2 related problems, both traceable to the fact that Liverpool allows you to play on others sets and groups.
1. played_cards in HandAndFoot can be totally local, but it cannot be local in Liverpool.
2. indexing scheme needs to be consistent throughout -- instead of keying on card.number, key on tuple = player#, group#.   
After some false starts those issues are resolved.
--
Changes in this commit pass more variables from RunClient to handView. 


Now merge visible cards which has a dictionary for each player to one big dictionary, so can play on existing groups. 
 Note this  made  compressGroups in tableView more complicated, but will simplifies other stuff.


* visible_cards needed definition modification in Liverpool; 
In Liverpool (but not HandAndFoot) changed structure of visible_cards so that it is now a list containing a single dictionary, instead of a dictionary for each player.  That single dictionary is the the one with the most cards played of all the dictionaries received from the player channels.  This allows you to preserve cards played by a player who later drops out. AND to preserve the ordering of runs, which should simplify assigning wilds.
-- 
Created method proccessRuns  in liverpool.py.  Puts run in order and  does some of the rule checking.  

Significant changes-- Cards now have property card.tempnumber so that can track assigned values of wilds, and when Aces are high. Server code does NOT use this property, as it is not passed between players and server.

processRuns is a new method that puts cards in order (including unassigned jokers). It simultaneously does rules checking.

* Oct28 cleanup (#9)

This now supports Liverpool with some caveats.  Cannot yet buy cards, and cannot designate where Aces and wilds should be when they're at the ends of runs.   Also this needs more polishing., and more thorough testing.

You can play on other players melds.  To support this needed to send visible cards to controller. Changed structure so that visible cards is now a list containing at most one dictionary, and the keys are tuples (player no. , group no. ).  There is a button for each group which is used when preparing cards.
Mechanics a bit different because you designate the group cards go to explicitly upon preparing them (each group has a different button).

I double checked that pyinstaller works.

* Buy option (#10)

Concluded buyOption had to be controlled by server.
Got server to send announcement to eligible player that they could buy top card.
Server now lets eligible clients know when buying opportunity occurs.
Clients let server know if they want the card.    Gives purchased card and top card to player who buys it.
Reset want_card so that don't accidentally purchase future discard if wanted earlier one.
--Other new features:
 Suit Symbols
changed font so In Liverpool: Tableview now displaying suit symbols.
Players can join between rounds  in Liverpool.
Moved message that appears at start and between rounds to ruleset (both HandAndFoot and Liverpool).

Moved processRuns outside of ruleset and into RunManagement.py.  
Updated structure so that now have controller handling prepared cards (good).   

* began work on wildsHiLo method, but decided to put that in new Branch, so commented out that work here.
Plan is to make branch off of this focused on wildHiLo, and then get BuyOption working with ambiguous wilds handled  automatically (by making them high).  Can come back to wildHILo work later.

Confirmed BuyOption working, pyinstaller working, HandAndFoot still working.


* Enhanced Tableview display in Liverpool
* pick-ups announced
             If you pick up the pile (HandAndFoot) or the top card (Liverpool)it is now announced.
* Can now start at any round !
           Done on server side. To start now need:
           py RunServer.py server:port Ruleset  # where # = starting round.
Features that were not in Liverpool branch that was just approved for merger into this fork's master:
* refresh draw pile
* cleaned up sortkeys code
* Update TableView.py (replaced 11,12,13,1 with J,Q,K,A in display of runs in Liverpool).
* fixed some bugs in RunManagement
* fixed bug in GamerServer that  caused crashes

Addressed all but 1 comment (tableView imports from rules)  from earlier review of Liverpool branch.

put check on length of visible_cards  (to insure no cards lost) in a separate method and now only call it on update, instead of on every broadcast

* Player disconnect liverpool (#27)

If player disconnects Liverpool should recover -- added methods to GameServer to handle this.
Checked that compiled, and that both HandAndFoot and Liverpool are working.  Removed print statements used in debugging.

Also cmade some cosmetic changes in response to user feedback:
Joker not offered for sale at beginning of game
Liverpool sorting buttons are a little bit more separated.
Some messages around card buying have been changed.

* Update RunClient.py

Give user 2nd chance to enter correct name of game.

* Update RunClient.py

* Reconciling.

* Server pick game (#29)

Main update is that now the client gets the name of the ruleset from the server.  This required changes to RunClient.py,
Controller.py, Clientstate.py,  GameServer.py and PlayerChannel.py.

Minor change:  hostinfo and port now entered separately.
Did this so that players would not need to type in host name when we switched to another port#, but could continue to use default.  

Also verified the if/while statements used to set player_index in RunClient.py are needed (while loop goes through a couple of times).

* Liverpool improvements (#30)

* Rules enforce last round is pat hand.
* Acknowledge response to buying opportunity.
* Wrap groups
                        For Liverpool with lots of players the text in groups needs to wrap around or you can't see the entire run.

- Verified that RunClient.py compiles.
- Removed some errant print statements.
- Added more info to print statements in RunClient.

Created RunClientWithJanServer -- a version that runs with the last releases server (does not get game  choice from server).

Co-authored-by: Helen <hwoodward@users.noreply.github.com>

slwoodward committed:
Liverpool improvements (#31)

* Rules enforce last round is pat hand.

* Acknowledge response to buying opportunity.

* Wrap groups

For Liverpool with lots of players the text in groups needs to wrap around or you can't see the entire run.

* Tested:
- Verified that RunClient.py compiles.
- Created version of RunClient that will run with previous version of server software that's on a remote server.
- Removed some errant print statements.
- Added more info to print statement in RunClient.

* Created a 2nd client executable: RunClientWithJanServer -- a version that runs with the last releases server (does not get game  choice from server).

slwoodward committed:
playerLeftGame (#32)

* playerLeftGame

added method in HandView to handle moving buttons over when a player drops out of a game mid-round.   Buttons spaced appropriately and prepared cards are cleared.  
Also moved call to "self.controller.resetProcessedCards(self.visible_scards)" into this method.

slwoodward committed:
Hopefully final change to this fork.  (#33)

* added method in HandView to handle moving buttons over when a player drops out of a game mid-round.   Buttons spaced appropriately and prepared cards are cleared.  Also moved call to "self.controller.resetProcessedCards(self.visible_scards)" into this method.

* fixed turn_index when player drops, else the next player might lose their turn.

* Cosmetic

slwoodward committed:
Merge branch 'master' into HopeForkDone

slwoodward committed:
Delete RunClientWithJanServer.py

slwoodward committed:
Merge branch 'master' into HopeForkDone

hwoodward commented on source/client/ClientState.py:
```diff
@@ -8,22 +8,30 @@ class ClientState:
-            #This is the unit test case - we may want to put a dummy ruleset in
-            print("In unittest mode - using HandAndFoot rules")
-            rule_module = "common.HandAndFoot"
+        self.ruleset = ruleset
```
This was a pre-existing thing, but we probably don't need to keep self.ruleset or require it on init, since we are passing it to the importRuleset method.

slwoodward replied:
With current architecture, Tableview needs to be passed the ruleset.  
I was able to eliminate clientState.rule_module (rule_module is now purely local to importRuleset method).

slwoodward replied:
I did eliminate the duplicate code.

hwoodward commented on source/client/ClientState.py:
```diff
@@ -8,22 +8,30 @@ class ClientState:
+            rule_module = "common.HandAndFoot"
+        self.rules = importlib.import_module(rule_module)
+
+    def getPlayerIndex(self, player_names):
```
This is duplicate code, please remove it

slwoodward replied:
Done.

hwoodward commented on source/client/Controller.py:
```diff
@@ -413,6 +415,14 @@ def Network_connectionDenied(self, data):
     ### Gameplay messages ###
+
+    def Network_defineGame(self, data):
+        #todo: is this if statement necessary?
```
You could just call the import from here instead of calling it in the runClient loop as it is now.

hwoodward replied:
The if is not REQUIRED, but a check you have the right amount of info is never WRONG, just remember to raise a clear exception in the else case so it is obvious what happened if you ever hit it.

slwoodward replied:
Moved import as suggested.   Replaced existing print statement with an exception.

hwoodward commented on source/server/GameServer.py:
```diff
@@ -184,15 +187,13 @@ def Send_publicInfo(self):
         if self.rules.Shared_Board:
             # Shared_Board is True: (e.g. Liverpool) -- each player transmits entire board of visible_cards to server.
-            self.Send_broadcast({"action": "publicInfo", "player_names": [p.name for p in self.players],"visible_cards": [self.visible_cards_now],"hand_status": [p.hand_status for p in self.players]})
+            self.Send_broadcast({"action": "publicInfo", "player_names": [p.name for p in self.players],"visible_cards": [self.visible_cards_now],"hand_status": [p.hand_status for p in self.players], "ruleset": self.ruleset})
```
is the ruleset actually USED from the broadcast? THere is an explicit call to request it, indpendent of this.

slwoodward replied:
Removed it from broadcast, and ran program to insure it wasn't used somewhere non-obvious.

hwoodward reviewed (changes requested):
You have duplicate code, and there are some improvements you could make to how you are sending and setting the ruleset from the server to the client.

slwoodward committed:
Addressing reviewer comments

slwoodward commented:
Addressed comments.  
 - With current architecture, TableView module uses self._state.ruleset.   Will be able to improve handling of ruleset in next version.
Also verified that previous version of RunClient will work with this server.
